Feat/weather response

zackerms opened the pull request:
## 実装の背景・目的

利用者のふとした行動に応じて、利用者の健康を支援するために機能を追加した。

利用者が天気を調べたときに、その日の天気と一緒に、
天気に応じた簡単な運動を促すメッセージを送信する機能を実装した。

## やったこと

- 利用者に天気を聞かれたときに、その日の天気を調べる。
- 外出できる場合と、そうでない場合で提示するアドバイスを変える。
  - 外出できるかどうかを判定するメソッドに対して、単体テストを追加した。 
- 内部エラーに応じたメッセージの送信
  - APIリクエストが上限に達したときや、APIリクエストに失敗したときに、ユーザーにその旨を伝える 

## キャプチャ

<img width="707" alt="スクリーンショット 2021-11-17 18 32 24" src="https://user-images.githubusercontent.com/55840281/142174651-73b8a7d6-2d27-4628-9790-c1bd95b6e482.png">


## 補足

zackerms committed:
feat(model): implemented can go out method

外に出られるかを調べてから、利用者にどのような運動を促すかを決める機能を実装するため、
HourlyWeather#can_go_outというメソッドを追加した。

zackerms committed:
feat(model): added japanese weather description

APIから英語で取得される天気を日本語で簡単に表現するメソッドを追加した。

APIのパラメーターで日本語を指定することができるが、「薄い雲」といった
利用者の期待しない天気が得られるため、このメソッドを追加した。

zackerms committed:
fix(model): moved exception file

ファイルの位置が、moduleの構造に従うように移動した。

zackerms committed:
build(app): added exception directory

Railsアプリケーションがapp/exceptionディレクトリを認識できるように、
設定に追加した。

zackerms committed:
feat(controller): implemented weather response

利用者が天気を調べたときに、その日の天気予報と
天気に応じた簡単な運動を促す機能を追加した。

newrice commented on app/controllers/webhook_controller.rb:
```diff
@@ -29,6 +29,9 @@ def callback
             advice_response(client, event)
+          elsif user_message.include?("天気")
+            # メッセージに「天気」が含まれている場合は、今日の天気と、おすすめの運動を提案する
+            weather_response(client, event)
```
以前、eguchi さんが指摘してくれた「天気」かつ「アドバイス」が含まれていた場合は？の問題がありそう。
「天気」が含まれている以上、天気が知りたそうな感じはあるかな？どうだろ

egurinko replied:
アドバイスの天気はないけど、天気のアドバイスがあると考えると天気を先にしたほうが良いかもですね 👀 

とはいえ、このまま単純な includes を使いながらハンドリングするメッセージを増やしてしまうと、どうしても順番が意図通りにできなくなるので、最終的には別のアプローチが必要そう。例えば、特定の文言が含まれているかどうかだけではなく、文言の順番などを正規表現などを使いつつやる方法。ただしめっちゃテクい。

もう一つは、includes 等を使わずに完全一致に倒ししてしまい、入力される文言をリッチメニューなどで制御する方法です。

とはいえ、主眼ではないので優先順位的に落としちゃうというアプローチはありそうです。

zackerms replied:
２つのものは同時には来ないだろうと思い、順番は気にしていませんでしたが、
確かに、BOTから天気に応じたアドバイスを貰った利用者は、次に「天気」と「アドバイス」をメッセージに含める可能性がありますね。

完全に利用者の意図に合わせた返答をする処理を書くのには時間がかかると思うので、
修正はマッチングの順番を入れ替えるのみにします！

newrice commented on app/models/api/open_weather_map/hourly_weather.rb:
```diff
@@ -16,6 +18,30 @@ def self.from_json(json)
+
+      # 外出できる天気かどうかを判定する。
+      def can_go_out?
+        [Weathers::CLEAR, Weathers::CLOUDS].include?(weather)
```
インスタンス変数にアクセスしていると思うので `@weather` で良さそう

newrice commented on test/models/api/open_weather_map/hourly_weather_test.rb:
```diff
@@ -0,0 +1,21 @@
+require 'test_helper'
+require_relative '../../../../app/models/api/open_weather_map/weathers'
```
なくても良さそう

newrice commented on test/models/api/open_weather_map/hourly_weather_test.rb:
```diff
@@ -0,0 +1,21 @@
+  module OpenWeatherMap
+    class HourlyWeatherTest < ActiveSupport::TestCase
+
+      def test_can_go_out
```
分岐をテストしたい感じ素敵です！

あまり本質じゃないので参考までですが、、、
ここでは、can_go_out の判定が正しいか
を確認すると思うので、想定しているケース+想定外文字列をテストしたほうが安全ではあるかなと思いました。
さらに参考までですが、↑のようにすると、同じコードが増えますのでテーブルドリブンテストのように書けそうかなと思いました。

```ruby
      def test_can_go_out
        [
          [Weathers::THUNDERSTORM, false],
          # and more...
          ["hogehoge", false],
        ].each do |input, expected|
          weather = HourlyWeather.new(time_unix: 0, temperature: 0, weather: input)
          assert weather.can_go_out? == expected
        end
      end
```

zackerms replied:
テストケースを先に書いていくのは良さそうです！

APIの仕様変更がなければ大丈夫なのですが、`Cloud`を`cloud`と小文字で渡した場合に
弾くようにもテストを修正したいと思います。

newrice commented on app/controllers/webhook_controller.rb:
```diff
@@ -89,4 +92,38 @@ def advice_response(client, event)
+      return
+    rescue => e
+      Rails.logger.error("[WebhookController] Error while fetching weather forecast: #{e}")
+      message = create_text_object(" ごめん、天気を調べられなかったよ。 \n時間をあけてもう一度聞いてね。 ")
```
文頭にスペースが入っている？

zackerms replied:
入っていました！
細かいところまでチェックして頂き、ありがとうございます！

newrice commented on app/controllers/webhook_controller.rb:
```diff
@@ -89,4 +92,38 @@ def advice_response(client, event)
+
+    # 天気を報告
+    forecast_three_hours_later = forecasts[0]
+    weather_japanese = forecast_three_hours_later.weather_japanese
```
ただの参考ですが、国際化はこんなのがあるみたいですね
https://railsguides.jp/i18n.html

egurinko commented on app/controllers/webhook_controller.rb:
```diff
@@ -89,4 +92,38 @@ def advice_response(client, event)
+    # 例外が発生した場合は、そのことを利用者に伝えて終了する。
+    begin
+      forecasts = api_client.fetch_three_hourly_forecasts
+    rescue API::TooManyRequestsException => e
```
うまくハンドリングできていそうです 🙋

egurinko commented on test/models/api/open_weather_map/hourly_weather_test.rb:
```diff
@@ -0,0 +1,21 @@
+require 'test_helper'
+require_relative '../../../../app/models/api/open_weather_map/weathers'
+
+module API
```
テストまで書いてるの素晴らしい

zackerms committed:
fix(controller): modified message pattern match

天気をBOTに聞いて、アドバイスを受け取った利用者が
次に天気に応じたアドバイスをもらおうとしたときに
適切なレスポンスが変えるように修正をした。

例：「今日の天気はどうなりそう？　どんなことをすればいいか、アドバイスをちょうだい」

zackerms committed:
refactor(model): use instance variable

zackerms committed:
test(model): added unexpected input

想定されている以外の天気が指定された場合の
テストケースを追加した。

zackerms committed:
refactor(model): removed unnecessary require statement

モジュールの名前解決のために一時的に利用していた、
require の宣言を削除した。

zackerms committed:
fix(controller): removed space in message

egurinko reviewed (approved):
概ね実装良さそうです 👌
